packages/gatsby/src/utils/websocket-manager.ts

```diff
@@ -25,87 +25,74 @@ type PageResultsMap = Map<string, IPageQueryResult>
         publicDir,
         pagePath
       )
 
-      return {
-        result: pageData,
-        id: pagePath,
-      }
+      result.result = pageData
     } catch (err) {
       throw new Error(
         `Error loading a result for the page query in "${pagePath}". Query was not run and no cached result was found.`
       )
     }
   }
 
-  return {
-    id: pagePath,
-    result: undefined,
-  }
+  return result
 }
 
-const hashPaths = (
-  paths?: Array<string>
-): undefined | Array<string | undefined> => {
-  if (!paths) {
-    return undefined
+/**
+ * Get page query result for given page path.
+ * @param {string} pagePath Path to a page.
+ */
+async function getStaticQueryData(
+  staticQueryId: string
+): Promise<IStaticQueryResult> {
+  const { program } = store.getState()
+  const publicDir = path.join(program.directory, `public`)
+
+  const filePath = path.join(
+    publicDir,
+    `page-data`,
+    `sq`,
+    `d`,
+    `${staticQueryId}.json`
+  )
+
+  const result: IStaticQueryResult = {
+    id: staticQueryId,
+    result: undefined,
   }
-  return paths.map(path => {
-    if (!path) {
-      return undefined
+  if (await fs.pathExists(filePath)) {
+    try {
+      const fileResult = await fs.readJson(filePath)
+
+      result.result = fileResult
+    } catch (err) {
+      // ingore errors
     }
-    return createHash(`sha256`).update(path).digest(`hex`)
-  })
+  }
+
+  return result
 }
 
-/**
- * Get cached StaticQuery results for components that Gatsby didn't run query yet.
- * @param {QueryResultsMap} resultsMap Already stored results for queries that don't need to be read from files.
- * @param {string} directory Root directory of current project.
- */
-const getCachedStaticQueryResults = (
-  resultsMap: QueryResultsMap,
-  directory: string
-): QueryResultsMap => {
-  const cachedStaticQueryResults: QueryResultsMap = new Map()
-  const { staticQueryComponents } = store.getState()
-  staticQueryComponents.forEach(staticQueryComponent => {
-    // Don't read from file if results were already passed from query runner
-    if (resultsMap.has(staticQueryComponent.hash)) return
-    const filePath = path.join(
-      directory,
-      `public`,
-      `page-data`,
-      `sq`,
-      `d`,
-      `${staticQueryComponent.hash}.json`
-    )
-    const fileResult = fs.readFileSync(filePath, `utf-8`)
-    if (fileResult === `undefined`) {
-      console.log(
-        `Error loading a result for the StaticQuery in "${staticQueryComponent.componentPath}". Query was not run and no cached result was found.`
-      )
-      return
-    }
-    cachedStaticQueryResults.set(staticQueryComponent.hash, {
-      result: JSON.parse(fileResult),
-      id: staticQueryComponent.hash,
-    })
-  })
-  return cachedStaticQueryResults
+function hashPaths(paths: Array<string>): undefined | Array<string> {
+  return paths
+    .filter(Boolean)
+    .map(path => createHash(`sha256`).update(path).digest(`hex`))
```

Comment:
Old code considered the `path` to be nullable (or maybe empty string? because type was `Array<string>`). New code does not make this exception. If the `!path` part was indeed relevant to check then not doing this anymore could lead to very subtle and hard to catch hash collisions.